pkg/networkservice/common/connect/server.go

```diff
@@ -34,27 +35,58 @@ import (
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 )
 
+// Contain grpc connection and client connection associated
+type clientEntry struct {
+	clientURI       *url.URL
+	client          networkservice.NetworkServiceClient
+	cc              grpc.ClientConnInterface
+	ready           chan bool // Flag for connection is establishing
+	connections     map[string]*networkservice.Connection
+	pending         int // A number of pending connections, in progress, waiting for connection establish, or performing request.
+	err             error
+	closeConnection func()
+}
+
+func (ce *clientEntry) markAsReady(state bool) {
+	ce.ready <- state
+	close(ce.ready)
+}
+
 type connectServer struct {
-	ctx                     context.Context
-	dialOptions             []grpc.DialOption
-	clientFactory           func(ctx context.Context, conn grpc.ClientConnInterface) networkservice.NetworkServiceClient
-	uRLToClientMap          clientMap // key == url as string
-	connectionIDToClientMap clientMap // key == connection.GetId()
-	dialOptionFactories     []func(ctx context.Context, request *networkservice.NetworkServiceRequest, clientURL *url.URL) []grpc.DialOption
+	ctx                 context.Context
+	dialOptions         []grpc.DialOption
+	clientFactory       func(ctx context.Context, conn grpc.ClientConnInterface) networkservice.NetworkServiceClient
+	clientMap           map[string]*clientEntry // key == url as string
+	connections         map[string]*clientEntry // Connection map is required to close using connection id.
+	dialOptionFactories []func(ctx context.Context, clientURL *url.URL) []grpc.DialOption
+	executor            serialize.Executor
+	dialer              func(ctx context.Context, target string, opts ...grpc.DialOption) (conn grpc.ClientConnInterface, cancelFunc func(), err error)
 }
 
 type dialOptionFactory struct {
```

Comment:
We need to find a way to use Callback without having to make odd code changes up and down our codebase (like DialOptFactory).

Reply:
It is not for callback but for testing, I think I could remove it and use grpc endpoint instead.

Reply:
Let's look into that.